File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:

        vals = []
        for l in lists:
            while l:
                vals.append(l.val)
                l = l.next
        vals.sort()
        dummy = ListNode()
        tmp = dummy
        for val in vals:
            tmp.next = ListNode(val)
            tmp = tmp.next
        return dummy.next
        
        

# key points
# 1. merge k linked-lists
# 2. 0 <= lists[i].length <= 500
# 3. 0 <= k <= 104
# The sum of lists[i].length will not exceed 104.

# 1. get length
# 2. get first value from first node of all lists
# 3. compare all first values and list up

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        std::vector<int> vals;
        
        for(auto ls : lists) {
            ListNode* sub = ls;
            while(sub){
                vals.push_back(sub->val);
                sub = sub->next;
            }
        }

        sort(vals.begin(), vals.end());
        ListNode dummy;
        ListNode *tmp = &dummy;
        for(auto val : vals) {
            tmp->next = new ListNode(val);
            tmp = tmp->next;
        }
        return dummy.next;
    }
};